doc/ipmiutil.8
.TH IPMIUTIL 8 "Version 1.1: 10 May 2012"
.SH NAME
ipmiutil \- a meta-command to invoke various IPMI functions.

.SH SYNOPSIS
.B "ipmiutil <command> [-x -NUPREFJTVY] [other command options]

.SH DESCRIPTION
This utility performs various IPMI functions.   Each of the individual
commands in the ipmiutil project can be invoked via this meta-command.
The <command> is one of the following:
   alarms   show/set the front panel alarm LEDs and relays
   leds     show/set the front panel alarm LEDs and relays
   cmd      send a specified raw IPMI command to the BMC
   config   list/save/restore BMC configuration parameters
   dcmi     get/set DCMI parameters
   discover    discover all IPMI servers on this LAN
   ekanalyzer  run FRU-EKeying analyzer on FRU files (deprecated, see fru)
   events   decode IPMI events and display them
   firewall show/set firmware firewall functions
   fru      show decoded FRU inventory data, write asset tag
   fwum     OEM firmware update manager extensions
   getevt   get IPMI events and display them, event daemon
   getevent get IPMI events and display them, event daemon
   health   check and show the basic health of the IPMI BMC
   hpm      HPM firmware update manager extensions
   lan      show/set IPMI LAN parameters and PEF table
   picmg    show/set picmg extended functions
   reset    cause the BMC to reset or power down the system
   sel      show/clear firmware System Event Log records
   sensor   show Sensor Data Records, readings, thresholds
   serial   show/set IPMI Serial & Terminal Mode parameters
   sol      start/stop an SOL console session
   smcoem   SuperMicro OEM functions
   sunoem   Sun OEM functions
   delloem  Dell OEM functions
   tsol     Tyan SOL console start/stop session
   wdt      show/set/reset the watchdog timer
.br
For help on each command (e.g. 'sel'), enter:
   ipmiutil sel -?
.br
For man pages on each command, its man page is named "i<command>", 
or refer to SEE ALSO below.

.br
This utility can use either the /dev/ipmi0 driver from OpenIPMI,
the /dev/imb driver from Intel, the /dev/ipmikcs driver from valinux,
direct user-space IOs, or the IPMI LAN interface if \-N is used.  

.SH OPTIONS
Command options are described in the man page for each command.
Below are a few of the common options.
.IP "-x"
Causes extra debug messages to be displayed.
.IP "-N nodename"
Nodename or IP address of the remote target system.  If a nodename is
specified, IPMI LAN interface is used.  Otherwise the local system
management interface is used.
.IP "-U rmt_user"
Remote username for the nodename given.  The default is a null username.
.IP "-P/-R rmt_pswd"
Remote password for the nodename given.  The default is a null password.
.IP "-E"
Use the remote password from Environment variable IPMI_PASSWORD.
.IP "-F drv_t"
Force the driver type to one of the followng:  
imb, va, open, gnu, landesk, lan, lan2, lan2i, kcs, smb.
Note that lan2i means lan2 with intelplus.
The default is to detect any available driver type and use it.
.IP "-J"
Use the specified LanPlus cipher suite (0 thru 17): 0=none/none/none,
1=sha1/none/none, 2=sha1/sha1/none, 3=sha1/sha1/cbc128, 4=sha1/sha1/xrc4_128,
5=sha1/sha1/xrc4_40, 6=md5/none/none, ... 14=md5/md5/xrc4_40.
Default is 3.
.IP "-T"
Use a specified IPMI LAN Authentication Type: 0=None, 1=MD2, 2=MD5, 4=Straight Password, 5=OEM.
.IP "-V"
Use a specified IPMI LAN privilege level. 1=Callback level, 2=User level, 3=Operator level, 4=Administrator level (default), 5=OEM level.
.IP "-Y"
Yes, do prompt the user for the IPMI LAN remote password.
Alternatives for the password are \-E or \-P.

.SH "EXAMPLES"
ipmiutil sel 
.br
Shows the IPMI System Event Log entries.
.PP
ipmiutil wdt 
.br
Shows the watchdog timer values.
.PP

.SH "SEE ALSO"
ialarms(8) iconfig(8) icmd(8) idiscover(8) ievents(8) ifirewall(8) 
ifru(8) ifruset(8) ifwum(8) igetevent(8) ihealth(8) ihpm(8) 
ilan(8) ipicmg(8) ireset(8) isel(8) isensor(8) iserial(8) isol(8) isunoem(8) 
iwdt(8) ipmiutil(8) ipmi_port(8)

.SH WARNINGS
See http://ipmiutil.sourceforge.net/ for the latest version of ipmiutil and any bug fix list. 

.SH COPYRIGHT
Copyright (C) 2009  Kontron America, Inc.
.PP
See the file COPYING in the distribution for more details
regarding redistribution.
.PP
This utility is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY.

.SH AUTHOR
.PP
Andy Cress <arcress at users.sourceforge.net>
.br



NEWS
IPMIUTIL NEWS 

In October 2001, this project started as part of the Carrier Grade Linux effort.
It was then known as 'panicsel' ("http://panicsel.sf.net") and included a kernel patch to write a Linux panic event to the IPMI firmware log, as well as utilities.  

In July 2002, this code was first included in MontaVista CGE Linux 2.1.
The panicsel functionality was included in OSDL CGL 1.0 and 2.0 requirements.
The kernel panic functionality was included in the OpenIPMI driver for 2.6 Linux kernels, and in the Intel IMB Linux driver v28 and greater.

A Linux rpm and a Windows setup package for ipmiutil is included on 
the platform CD with Intel carrier-grade servers.

The ipmiutil (or panicsel) rpm is known to be included in the following 
Linux distributions:
MontaVista CGE 2.1, 3.0, 3.1, 4.0, SuSE SLES9, SLES10, RedFlag 5.0

See http://ipmiutil.sourceforge.net for binaries, documentation, etc.

In Oct 2001 (0.8.1), started panicsel.
In Dec 2001 (1.0.0), released with Intel Telecom Linux Technology project
In Jul 2002 (1.2.1), released with MontaVista CGE 2.1
In Jan 2004 (1.4.1), added support for Windows 2000 & 2003
In Aug 2004 (1.4.8), the project was moved from panicsel.sf.net to ipmiutil.sf.net.  The new name more clearly reflects the purpose of the project in its current state.  
In Nov 2004 (1.5.0), support for IPMI LAN was added using FreeIPMI.
In May 2006 (1.7.0), built-in IPMI LAN support was added.
In Aug 2006 (1.7.5), driverless direct I/O support was added
In Jan 2007 (1.8.0), added ipmiutil meta-command, IPMI LAN 2.0 (lanplus) support
In Feb 2007 (1.9.2), added Linux Serial-Over-Lan console feature 
In Oct 2007 (2.0.0), added Windows SOL console and remote soft-shutdown features
In Mar 2008 (2.1.0), the Windows SOL client application was completed.
In Aug 2008 (2.2.0), iconfig function was added to save/restore the BMC configuration parameters
In Sep 2008 (2.2.1), support for the Microsoft IPMI driver was added.
In Nov 2008 (2.3.0), support for the Sun Solaris bmc driver was added.
In Mar 2009 (2.3.5), support for the FreeBSD OS was added.
In Oct 2009 (2.5.0), added ANSI VT100  emulation for Windows SOL console
In Feb 2010 (2.6.0), new subcommand naming scheme (i+subcmd)
In Apr 2010 (2.6.2), ifruset was added to set all FRU Product Area fields
In Jun 2010 (2.6.5), added picmg and firewall functionality
In Aug 2010 (2.6.8), Fedora 14 version of ipmiutil submitted
In Sep 2010 (2.6.9), added fwum/hpm functionality, and OEM Fujitsu sensor/sel 
In Sep 2010 (2.7.0), added sunoen, ekanalyzer, fru_picmg, oem_kontron
In Sep 2010 (2.7.1), added ppc64, mips64, sparc arch support
In Nov 2010 (2.7.2), added Debian package support
In Nov 2010 Fedora 14 released with ipmiutil-2.6.8
In Dec 2010 (2.7.3), added SuperMicro OEM functions
In Mar 2011 OpenSuSE 11.4 released with ipmiutil-2.7.2
In May 2011 (2.7.6), added VLAN support and WIN64 fixes
In Sep 2011 (2.7.8), added IPv6, systemd, Dell OEM functions, PICMG 2.3, tsol
In Oct 2011 (2.7.9), support Wind River and ARM cross-compile, Intel S2600
In Dec 2011 (2.8.0), added DCMI 1.1 commands, devel pkg, MSI package
In Feb 2012 (2.8.1), added Quanta S99Q functions, Dell fixes, Intel ME fixes
In Mar 2012 (2.8.2), added Intel Romley OEM support
In Jun 2012 (2.8.4), added support for HP-UX builds and DCMI 1.1 commands
In Oct 2012 (2.8.6), added pre-built Windows DLL for sample development apps
In Feb 2013 (2.8.8), added much more SuperMicro OEM support
In Apr 2013 (2.9.0), added iseltime to sync sys/RTC/SEL times
In May 2013 (2.9.1), added support for MacOSX client
In Dec 2016 (3.0.1), added support for Lenovo OEM sensors
In Feb 2017 (3.0.1), added support for ASUS OEM sensors
In Mar 2017 (3.0.3), added iuser to consolidate user functions



util/ipmiutil.c
/***********************************************
 * ipmiutil.c 
 *
 * This is a meta-command utility to invoke each of the
 * other sub-commands in a consolidated interface.
 * To build this, compile with -DMETACOMMAND.
 *
 * Author:  Andy Cress  arcress at users.sourceforge.net
 * Copyright (c) 2006-2007 Intel Corporation.
 * Copyright (c) 2009 Kontron America, Inc.
 *
 * 01/03/07 ARCress - created
 * 01/05/07 ARCress - version 1.0
 * 01/10/07 ARCress - version 1.1
 * 02/07/07 ARCress - version 1.3 adding isolconsole
 * 02/26/07 ARCress - updated sub-command names
 * 08/31/07 ARCress - added "leds" subcommand
 *
 ***********************************************/
/*----------------------------------------------------------------------*
The BSD License 

Copyright (c) 2007, Intel Corporation
All rights reserved.

Redistribution and use in source and binary forms, with or without 
modification, are permitted provided that the following conditions are met:

  a.. Redistributions of source code must retain the above copyright notice, 
      this list of conditions and the following disclaimer. 
  b.. Redistributions in binary form must reproduce the above copyright notice,
      this list of conditions and the following disclaimer in the documentation 
      and/or other materials provided with the distribution. 
  c.. Neither the name of Intel Corporation nor the names of its contributors 
      may be used to endorse or promote products derived from this software 
      without specific prior written permission. 

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND 
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED 
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR 
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES 
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON 
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT 
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS 
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *----------------------------------------------------------------------*/
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>

#include "ipmicmd.h"
#include "ipmiutil.h"

static char *progname  = "ipmiutil";
char *progver   = "3.16";
// static char fdebug = 0;
/*int ipmiutil(int argc, char **argv); */

#define NSUBCMDS  29
static struct {
  int idx;
  char tag[16];
  int (*rtn)(int argc, char **argv);
  char desc[64];
  } subcmds[NSUBCMDS] = {
 {  0, "alarms", i_alarms,  "show/set the front panel alarm LEDs and relays" },
 {  1, "leds",   i_alarms,  "show/set the front panel alarm LEDs and relays" },
 {  2, "discover", i_discover, "discover all IPMI servers on this LAN" },
 {  3, "cmd",    i_cmd ,    "send a specified raw IPMI command to the BMC" },
 {  4, "config", i_config,  "list/save/restore BMC configuration parameters" },
 { 26, "dcmi",   i_dcmi,    "get/set DCMI parameters" },
 {  5, "ekanalyzer", i_ekanalyzer, "run EKeying analyzer on FRU files (deprecated, see fru)" },
 {  6, "events", i_events,  "decode IPMI events and display them" },
 {  7, "firewall",   i_firewall, "show/set firmware firewall functions" },
 {  8, "fru",    i_fru,     "show decoded FRU inventory data, write asset tag"},
 {  9, "fwum",   i_fwum,    "OEM firmware update manager extensions" },
 { 10, "getevt", i_getevt,   "get IPMI events and display them, event daemon" },
 { 11, "getevent", i_getevt, "get IPMI events and display them, event daemon" },
 { 12, "health", i_health,  "check and show the basic health of the IPMI BMC"},
 { 13, "hpm",    i_hpm,     "HPM firmware update manager extensions" },
 { 14, "lan",    i_lan,     "show/set IPMI LAN parameters, users, PEF rules"},
 { 15, "picmg",  i_picmg,   "show/set picmg extended functions" },
 { 25, "power",  i_reset,   "issue IPMI reset or power control to the system"},
 { 16, "reset",  i_reset,   "issue IPMI reset or power control to the system"}, 
 { 17, "sel",    i_sel,     "show/clear firmware System Event Log records" },
 { 18, "sensor", i_sensor,  "show Sensor Data Records, readings, thresholds" },
 { 19, "serial", i_serial,  "show/set IPMI Serial & Terminal Mode parameters"},
 { 20, "sol",    i_sol,     "start/stop an SOL console session" },
 { 21, "smcoem", i_smcoem,  "SuperMicro OEM functions" },
 { 22, "sunoem", i_sunoem,  "Sun OEM functions" },
 { 23, "delloem",i_delloem, "Dell OEM functions" },
 { 24, "tsol",   i_tsol,    "Tyan SOL console start/stop session" },
 { 28, "user",   i_user,    "list or modify IPMI LAN users" },
 { 27, "wdt",    i_wdt,     "show/set/reset the watchdog timer" }
 };

static char usagemsg[] = "Usage: ipmiutil <command> [other options]\n"
                         "   where <command> is one of the following:\n";
static char helpmsg[]  = "For help on each command (e.g. 'sel'), enter:\n"
			 "   ipmiutil sel -?\n";

static void show_usage()
{
    int i;
    printf("%s", usagemsg);
    for (i=0; i<NSUBCMDS; i++)
       printf("\t%s\t%s\n",subcmds[i].tag,subcmds[i].desc);
    printf("   common IPMI LAN options:\n");
    print_lan_opt_usage(0);
    printf("%s", helpmsg);
}

#ifdef DOS
int i_discover(int argc, char **argv)
{
    printf("The discover function is not supported in DOS.\n");
    return(1);
}
#endif

#ifndef _WINDLL
/* omit main if compiled as a Windows DLL */
#ifdef WIN32
int __cdecl
#else
int
#endif
main(int argc, char **argv)
{
   int ret = 1;
   int i;
   char *psubcmd = "";
   
   if (argc < 2) {
      printf("%s ver %s\n", progname,progver); 
      show_usage();
      ret = ERR_USAGE;
      goto do_exit;
   }
#ifdef TEST_LOOP
   /* special subcommand processing loop for testing */
#ifdef WIN32
   while ( !(( _kbhit() ) &&  (_getch() == 'q')) )
#else
   while ( 1 )
#endif
   {
        for (i = 0; i < NSUBCMDS; i++)
        {
            if (strcmp(argv[1],subcmds[i].tag) == 0) {
                psubcmd = argv[1];
                argc--;
                argv++;
                ret = subcmds[i].rtn(argc,argv);

                argc++; argv--;  /*requeue the same subcmd*/
                os_usleep( 1, 0 ); /*sleep 1 sec*/
                break;
            }
        }
   }
#else
   for (i = 0; i < NSUBCMDS; i++)
   {
       if (strcmp(argv[1],subcmds[i].tag) == 0) {
          psubcmd = argv[1];
          argc--;
          argv++;
          ret = subcmds[i].rtn(argc,argv);
	  break;
       }
   }
#endif
   if (i >= NSUBCMDS) {
#ifdef LINUX
      if ((strcmp(argv[1],"svc") == 0) && (argc >= 3)) {
	 char mycmd[80];
	 char *pfunc;
	 char *psvc;
	 char fchkok;
	 /* undocumented: start a given service, only works locally */
	 psvc = "ipmi_port";
	 if (argc > 2) pfunc = argv[2];
	 else pfunc = "on";
	 ret = system("ls /sbin/chkconfig >/dev/null 2>&1");
	 if (ret == 0) fchkok = 1;
	 else fchkok = 0;
 	 if (strcmp(pfunc,"off") == 0) {
	    sprintf(mycmd,"service %s stop\n",psvc);
	    printf("%s\n",mycmd);
	    ret = system(mycmd);
	    if (fchkok) {
	       sprintf(mycmd,"/sbin/chkconfig --del %s\n",psvc);
	       printf("%s\n",mycmd);
	       ret = system(mycmd);
	    }
	 } else {
	    if (fchkok) {
		sprintf(mycmd,"/sbin/chkconfig --add %s\n",psvc);
		printf("%s\n",mycmd);
		ret = system(mycmd);
		sprintf(mycmd,"/sbin/chkconfig --level 345 %s on\n",psvc);
		// printf("%s\n",mycmd);
		ret = system(mycmd);
	    }
	    sprintf(mycmd,"service %s start\n",psvc);
	    printf("%s\n",mycmd);
	    ret = system(mycmd);
	 }
      } else 
#endif
      {
         printf("%s ver %s\n", progname,progver); 
         show_usage();
         ret = ERR_USAGE;
      }
   } 

do_exit:
   {
      char tag[30];
      sprintf(tag,"%s %s",progname,psubcmd);
      show_outcome(tag,ret); 
   }
   return(ret);
}
#endif

/*end ipmiutil.c*/
